### ISSUE TITLE
Conformer CTC dropping spoken numbers

### ISSUE THREAD

**Issue by itzsimpl** [NVIDIA/NeMo#3937]
We have successfully trained a Conformer CTC BPE L model (non-English) with a very good WER in the lower teens mark. There is one peculiar thing that is happening though. The model is dropping numbers, interestingly not always and mostly dates. Especially if more are spoken in succession as part of a sentence. I have found a similar issue/questionon regarding the Spanish language on Riva Forums https://forums.developer.nvidia.com/t/riva-with-asr-doesnt-recognize-numbers-on-my-audio/203244, but no explanation. 

It is true that the train dataset does not cover such specific cases, but we were under the impression that in general the model learns to recognise BPE tokens from the waveform, and no "internal LM" or context dependency between BPE tokens (or at max a short one). In any case we'd expect the model to output at least something, not miss entire sections with multiple consecutive words. An example of the worst case we are seeing (translated in English) would be:
- reference: "accepted on twelfth fifth twenty twenty and modified on seventh seventh twenty twenty-one"
- transcription: "accepted and modified"  

We also get different results depending if we run transcription with `transcribe_speech.py`, `transcribe_speech_parallel.py` or buffered inference. 

Any thoughts to what might be the source of the problem, or what can we do to resolve it?

**Comment by VahidooX** [comment]
One possible reason can be that you have dropped numbers from the text target of the training data and the model has learned to drop numbers. 
Do you have any numbers in your training data? 
Do you convert the numbers to their written format?

**Comment by itzsimpl** [comment]
We do not drop numbers from training data and all data was prepared in a written form. Furthermore, I did a check for the mentioned worst case. The dates do not appear in the train set, but all individual words (components of the dates in question) appear at least once in the train set. So if the model is not learning an "internal LM", which in this case would does not see successions of dates, dropping them seems utterly unintuitive, as the recogniser should recognise at least something. It does so on other occasions of dates, just not when there are multiple consecutive. 

The individual words (components of the dates in question) can be tokenised, so it is not a BPE issue, there are training examples with these words, so they must have been seen at least once, the audio is there, a clean waveform. What else?

**Comment by titu1994** [comment]
Hmm CTC models do have conditional independence assumption but I've not seen such an extreme case where it drops nearly 80% of the sentence. Possibly there may be audio issues though I doubt that, but could you check out a few samples using speech data explorer in Nemo ?

**Comment by itzsimpl** [comment]
I have transcribed the entire dataset and given a rough check, but nothing really popped out. There are of course occasional really bad WERs but mostly due to bad transcripts, segmentation issues or terrible audio, but that is to be expected. I know one thing I can do more is to clean up the dataset even further. However, when the proportion of such bad WERs is kept very low with respect to the entire dataset, as it is, it shouldn't have such a detrimental effect. Any other ideas?

**Comment by titu1994** [comment]
Ok let's take it a few sample at a time. Could you get maybe 6-8 samples where such significant issue occurs, and analyse the output of transcribe speech, transcribe parallel ? For now lets leave buffered for later. 

With those results let's do a sample by sample analysis of what exactly is being dropped.

**Comment by VahidooX** [comment]
It is not very likely for the model to drop words related to numbers or dates as even if there is no date/number in the train, then model should at least produce some close words (specially if it is a CTC model). If it happens repeatedly for many samples, the only reason come to my mind is that those patterns are dropped during the normalization/cleaning and model has forced to learn to drop and ignore those words.

**Comment by itzsimpl** [comment]
Thanks for the help. While preparing the samples we discovered a typo that lead to the inclusion of certain unwanted data in the train dataset, approx 3% of it. We're currently retraining the model with a cleaner dataset. We will recheck for the issue once the training is complete. Also for any differences between the different transcription scripts. 

I will leave the issue open, and update once we're done.

**Comment by itzsimpl** [comment]
We finished retraining the model, and it seems that the culprit for dropped words was in our preprocessing typo. We also rechecked for possible discrepancies between `transcribe_speech_parallel.py` and `transcribe_speech.py` and found out these are occasional, but on the level of single characters and I assume in the normal range of variability. Thank you once again for the help in tracking down the issue.
